core/shared_obj.h

```diff
@@ -71,18 +71,21 @@ class SharedObjectSpace {
  public:
   // By default, the Get() method creates a new object with its default
   // constructor (a constructor without any arguments). To change this behavior,
-  // you can use the optional argument "constructor", a callable object
+  // you can use the optional argument "creator", a callable object
   // (function pointer, functor, lambda function, etc.) that creates the object
   // in a way you'd like, e.g., to create an object with a non-default
   // constructor, or to reuse an already existing object.
+  //
+  // NOTE: the creator function will be called with the global mutex being held,
+  //       so it's not a good idea for the creator to block on something.
   template <typename T>
   std::shared_ptr<T> Get(
       const std::string &name,
-      std::function<std::shared_ptr<T>()> constructor = DefaultConstructor<T>) {
+      std::function<std::shared_ptr<T>()> creator = DefaultConstructor<T>) {
     SharedObjectKey key = std::make_pair(std::type_index(typeid(T)), name);
 
-    static std::mutex mutex;
-    std::lock_guard<std::mutex> lock(mutex);
+    static std::recursive_mutex mutex;
```

Comment:
Oh, I didn't notice that you were using a non-recursive mutex earlier!  This is better, one can make a shared object using another shared object in the same thread now. (Not sure we'll ever need this but it's generally inexpensive anyway, and I often think all mutexes should default to recursive.)